// @vibehub/core/dist/git-facade.js
/**
 * GitFacade — the ONLY place that shells out to `git` / `gh`
 * (decision-project-025 module list; decision-github-002: 团队可见通路 =
 * 纯 git + gh CLI,零服务端).
 *
 * Boundary rules implemented here:
 * - Worktrees resolve to their main repo via `git rev-parse --git-common-dir`
 *   (decision-github-004: N worktrees = 1 repo domain).
 * - `gh` absent or unauthenticated is an explicit degraded tier, not an
 *   error: pure-git facts keep flowing, PR facts come back as null
 *   (decision-github-004).
 * - Zero LLM, zero invention: every return value is a verbatim git/gh fact.
 */
import { spawnSync } from "node:child_process";
import crypto from "node:crypto";
import path from "node:path";
export class GitError extends Error {
    args;
    exitCode;
    stderr;
    constructor(args, exitCode, stderr) {
        super(`git ${args.join(" ")} failed (${exitCode}): ${stderr.trim()}`);
        this.args = args;
        this.exitCode = exitCode;
        this.stderr = stderr;
    }
}
function run(cmd, args, cwd) {
    const r = spawnSync(cmd, args, {
        cwd,
        encoding: "utf8",
        maxBuffer: 64 * 1024 * 1024,
    });
    return { status: r.status, stdout: r.stdout ?? "", stderr: r.stderr ?? "" };
}
function runBuffer(cmd, args, cwd) {
    const r = spawnSync(cmd, args, { cwd, encoding: "buffer", maxBuffer: 64 * 1024 * 1024 });
    return { status: r.status, stdout: r.stdout ?? Buffer.alloc(0), stderr: r.stderr?.toString("utf8") ?? "" };
}
const compareRepoPaths = (a, b) => Buffer.compare(Buffer.from(a, "utf8"), Buffer.from(b, "utf8"));
/** Parse an origin URL into "owner/name"; null when unparseable. */
export function parseRemoteSlug(url) {
    const m = url
        .trim()
        .match(/(?:[/:])([^/:]+)\/([^/]+?)(?:\.git)?\/?$/);
    return m ? `${m[1]}/${m[2]}` : null;
}
export class GitFacade {
    /** Main-repo working dir (worktrees already resolved to their main repo). */
    repoRoot;
    constructor(repoPath) {
        this.repoRoot = GitFacade.resolveRepoRoot(repoPath);
    }
    /**
     * decision-github-004: a path inside any worktree resolves to the MAIN
     * repo's root, so N worktrees land in one repo domain.
     */
    static resolveRepoRoot(anyPath) {
        const r = run("git", ["rev-parse", "--path-format=absolute", "--git-common-dir"], anyPath);
        if (r.status !== 0) {
            throw new GitError(["rev-parse", "--git-common-dir"], r.status, r.stderr);
        }
        return path.dirname(r.stdout.trim());
    }
    git(args) {
        const r = run("git", args, this.repoRoot);
        if (r.status !== 0)
            throw new GitError(args, r.status, r.stderr);
        return r.stdout;
    }
    tryGit(args) {
        const r = run("git", args, this.repoRoot);
        return r.status === 0 ? r.stdout : null;
    }
    /** "owner/name" from the origin remote; null when there is no remote. */
    remoteSlug() {
        const url = this.tryGit(["remote", "get-url", "origin"]);
        return url ? parseRemoteSlug(url) : null;
    }
    hasRemote() {
        return this.tryGit(["remote", "get-url", "origin"]) !== null;
    }
    /**
     * Default branch: origin/HEAD when known, else main/master if they exist
     * on the remote. Pure git fact — no invention.
     */
    defaultBranch() {
        const head = this.tryGit(["symbolic-ref", "refs/remotes/origin/HEAD"]);
        if (head)
            return head.trim().replace("refs/remotes/origin/", "");
        for (const cand of ["main", "master"]) {
            if (this.tryGit(["rev-parse", "--verify", `refs/remotes/origin/${cand}`]))
                return cand;
        }
        throw new GitError(["symbolic-ref", "refs/remotes/origin/HEAD"], 1, "cannot determine default branch (no origin/HEAD, no origin/main|master)");
    }
    /** `git fetch --prune origin`. Failure is a fact (offline), not a throw. */
    fetch() {
        const r = run("git", ["fetch", "--prune", "origin"], this.repoRoot);
        return { ok: r.status === 0, stderr: r.stderr };
    }
    /**
     * All remote branches except origin/HEAD, newest commit first. With
     * `aheadBehindVs` (a branch name on origin), each entry also carries
     * ahead/behind vs that branch in the SAME spawn (%(ahead-behind), git ≥
     * 2.41) — one call instead of 2 per branch; silently falls back to the
     * plain listing on older git.
     */
    listRemoteBranches(aheadBehindVs) {
        const base = "%(refname:short)%09%(objectname)%09%(committerdate:iso-strict)%09%(authorname)";
        let out = null;
        let withCounts = false;
        if (aheadBehindVs) {
            out = this.tryGit([
                "for-each-ref",
                "refs/remotes/origin",
                "--sort=-committerdate",
                `--format=${base}%09%(ahead-behind:origin/${aheadBehindVs})`,
            ]);
            withCounts = out !== null;
        }
        out ??= this.git([
            "for-each-ref",
            "refs/remotes/origin",
            "--sort=-committerdate",
            `--format=${base}`,
        ]);
        const branches = [];
        for (const line of out.split("\n")) {
            if (!line.trim())
                continue;
            const [ref, sha, date, author, counts] = line.split("\t");
            if (!ref || !sha || !date)
                continue;
            if (ref === "origin/HEAD" || ref === "origin")
                continue;
            const branch = {
                name: ref.replace(/^origin\//, ""),
                headSha: sha,
                lastCommitAt: date,
                lastAuthor: author ?? "",
            };
            if (withCounts && counts) {
                const [ahead, behind] = counts.trim().split(/\s+/).map(Number);
                if (Number.isFinite(ahead) && Number.isFinite(behind)) {
                    branch.ahead = ahead;
                    branch.behind = behind;
                }
            }
            branches.push(branch);
        }
        return branches;
    }
    /** True when `sha` is already contained in origin/<defaultBranch>. */
    isMerged(sha, defaultBranch) {
        const r = run("git", ["merge-base", "--is-ancestor", sha, `origin/${defaultBranch}`], this.repoRoot);
        return r.status === 0;
    }
    /** Commits ahead/behind of origin/<defaultBranch> (merge-base三点语义). */
    aheadBehind(branch, defaultBranch) {
        const out = this.git([
            "rev-list",
            "--left-right",
            "--count",
            `origin/${defaultBranch}...origin/${branch}`,
        ]);
        const [behind, ahead] = out.trim().split(/\s+/).map(Number);
        return { ahead: ahead ?? 0, behind: behind ?? 0 };
    }
    /**
     * The branch's footprint: files it changed vs the merge-base with the
     * default branch (`git diff --name-status A...B` = three-dot merge-base
     * semantics) — 队友足迹 diff (decision-github-002).
     */
    branchFiles(branch, defaultBranch) {
        const out = this.git([
            "diff",
            "--name-status",
            "--no-renames",
            `origin/${defaultBranch}...origin/${branch}`,
        ]);
        const files = [];
        for (const line of out.split("\n")) {
            if (!line.trim())
                continue;
            const [kind, ...rest] = line.split("\t");
            const p = rest[rest.length - 1];
            if (!kind || !p)
                continue;
            files.push({ path: p, changeKind: kind.charAt(0) });
        }
        return files;
    }
    /**
     * merge-tree conflict warning between two branches (decision-github-002:
     * 跨 branch 冲突预警,fetch 后全本地).
     *
     * Returns the conflicted paths, [] for a clean merge, or null when git
     * cannot simulate the merge at all (e.g. no common ancestor) — unknown is
     * reported as unknown, never as "clean".
     */
    mergeTreeConflicts(refA, refB) {
        const r = run("git", ["merge-tree", "--write-tree", "--name-only", refA, refB], this.repoRoot);
        if (r.status === 0)
            return [];
        if (r.status !== 1)
            return null;
        // Exit 1 = conflicts. Output: <tree-oid>\n<conflicted files...>\n\n<messages>
        const lines = r.stdout.split("\n");
        const paths = [];
        for (const line of lines.slice(1)) {
            if (line === "")
                break;
            paths.push(line);
        }
        return paths;
    }
    /**
     * Current branch AT a given path — deliberately static: the facade's own
     * commands run at the main-repo root (the repo DOMAIN), but a session in
     * a worktree has its own HEAD, which lives at the session's cwd.
     * Null on detached HEAD.
     */
    static currentBranchAt(anyPath) {
        const r = run("git", ["rev-parse", "--abbrev-ref", "HEAD"], anyPath);
        const name = r.status === 0 ? r.stdout.trim() : "";
        return name && name !== "HEAD" ? name : null;
    }
    /** Working-tree top level at a path (a worktree's own root, not the domain's). */
    static toplevelAt(anyPath) {
        const r = run("git", ["rev-parse", "--show-toplevel"], anyPath);
        return r.status === 0 ? r.stdout.trim() : null;
    }
    /**
     * The three session facts the hook path needs, in ONE spawn — `vibehub
     * hook` fires on every tool use and must stay milliseconds, so per-fact
     * subprocesses are the budget's biggest enemy. rev-parse emits the
     * answers line-by-line in argument order.
     */
    static sessionContextAt(anyPath) {
        const r = run("git", [
            "rev-parse",
            "--path-format=absolute",
            "--git-common-dir",
            "--show-toplevel",
            "--abbrev-ref",
            "HEAD",
        ], anyPath);
        if (r.status !== 0) {
            throw new GitError(["rev-parse", "(session context)"], r.status, r.stderr);
        }
        const [commonDir, toplevel, head] = r.stdout.trim().split("\n");
        if (!commonDir || !toplevel || !head) {
            throw new GitError(["rev-parse", "(session context)"], r.status, r.stdout);
        }
        return {
            repoRoot: path.dirname(commonDir),
            toplevel,
            branch: head !== "HEAD" ? head : null,
        };
    }
    /** Current commit at a session path; called once when a task is captured. */
    static headShaAt(anyPath) {
        const r = run("git", ["rev-parse", "--verify", "HEAD"], anyPath);
        if (r.status !== 0)
            throw new GitError(["rev-parse", "HEAD"], r.status, r.stderr);
        return r.stdout.trim();
    }
    /** True only when the exact object id names a commit reachable in this repo's object store. */
    static hasCommitAt(anyPath, commitSha) {
        if (!/^[0-9a-f]{40}$/.test(commitSha))
            return false;
        const r = run("git", ["cat-file", "-e", `${commitSha}^{commit}`], anyPath);
        return r.status === 0;
    }
    /** Exact tracked-tree denominator and delta between two resolved commits. */
    static commitInventory(anyPath, baseCommit, targetCommit) {
        const root = GitFacade.resolveRepoRoot(anyPath);
        for (const commit of [baseCommit, targetCommit])
            if (!GitFacade.hasCommitAt(root, commit))
                throw new GitError(["cat-file", commit], 1, "commit not found");
        const tree = runBuffer("git", ["ls-tree", "-r", "-z", targetCommit], root);
        if (tree.status !== 0)
            throw new GitError(["ls-tree", targetCommit], tree.status, tree.stderr);
        const targetEntries = tree.stdout.toString("utf8").split("\0").filter(Boolean).map(record => { const tab = record.indexOf("\t"); if (tab < 0)
            throw new GitError(["ls-tree", targetCommit], 1, "malformed tree entry"); const [mode, type, object] = record.slice(0, tab).split(" "); return { mode: mode, type: type, object: object, path: record.slice(tab + 1) }; }).sort((a, b) => compareRepoPaths(a.path, b.path));
        const diff = runBuffer("git", ["diff", "--name-status", "-z", "-M", baseCommit, targetCommit, "--"], root);
        if (diff.status !== 0)
            throw new GitError(["diff", baseCommit, targetCommit], diff.status, diff.stderr);
        const delta = new Map(), parts = diff.stdout.toString("utf8").split("\0").filter(Boolean);
        for (let i = 0; i < parts.length;) {
            const status = parts[i++];
            if (status.startsWith("R")) {
                const previousPath = parts[i++], nextPath = parts[i++];
                delta.set(nextPath, { changeKind: "renamed", previousPath });
            }
            else {
                const relative = parts[i++];
                delta.set(relative, { changeKind: status.startsWith("A") ? "added" : status.startsWith("D") ? "deleted" : "modified" });
            }
        }
        const rows = [];
        for (const entry of targetEntries) {
            const content = entry.type === "blob" ? runBuffer("git", ["cat-file", "blob", entry.object], root) : { status: 0, stdout: Buffer.from(entry.object, "utf8"), stderr: "" };
            if (content.status !== 0)
                throw new GitError(["cat-file", entry.object], content.status, content.stderr);
            rows.push({ path: entry.path, ...(delta.get(entry.path) ?? { changeKind: "unchanged" }), contentHash: crypto.createHash("sha256").update(content.stdout).digest("hex") });
        }
        for (const [relative, change] of delta)
            if (change.changeKind === "deleted")
                rows.push({ path: relative, ...change });
        return rows.sort((a, b) => compareRepoPaths(a.path, b.path));
    }
    /** Read-side commit events bounded by the task's captured HEAD. */
    commitEventsSince(startHeadSha, headRef) {
        const out = this.git([
            "log",
            "--reverse",
            "--format=%x1e%H%x1f%cI%x1f%s",
            "--name-only",
            `${startHeadSha}..${headRef}`,
        ]);
        return out
            .split("\x1e")
            .filter((block) => block.trim())
            .map((block) => {
            const lines = block.trim().split("\n");
            const [fullSha, at, message] = lines.shift().split("\x1f");
            if (!fullSha || !at || message === undefined) {
                throw new Error("malformed git log record");
            }
            const files = new Set(lines.map((line) => line.trim()).filter(Boolean));
            return {
                id: `git:${fullSha}`,
                at,
                type: "commit",
                sha: fullSha.slice(0, 7),
                message,
                filesChanged: files.size,
            };
        });
    }
    /**
     * Default branch, tolerant variant for hook ingestion: falls back to
     * "main" when there is no remote at all (a hook must never fail a
     * session over repo shape).
     */
    defaultBranchOr(fallback) {
        try {
            return this.defaultBranch();
        }
        catch {
            return fallback;
        }
    }
    /** Tracked-file count — the honest denominator for the unmapped gray. */
    lsFilesCount() {
        const out = this.git(["ls-files"]);
        return out.split("\n").filter((l) => l.length > 0).length;
    }
}
/* ── gh (PR facts) — explicit degraded tier when absent ─────────────────── */
export class GhFacade {
    repoRoot;
    constructor(repoRoot) {
        this.repoRoot = repoRoot;
    }
    /**
     * PR list via `gh pr list` (decision-github-002). Returns null when gh is
     * missing or unauthenticated — the explicit degraded tier of
     * decision-github-004: pure-git facts keep flowing, the UI labels the gap.
     */
    listPrs() {
        const r = run("gh", [
            "pr",
            "list",
            "--state",
            "all",
            "--limit",
            "200",
            "--json",
            "number,title,state,headRefName",
        ], this.repoRoot);
        if (r.status !== 0)
            return null;
        try {
            const rows = JSON.parse(r.stdout);
            return rows.map((p) => ({
                number: p.number,
                title: p.title,
                state: p.state.toLowerCase(),
                headRefName: p.headRefName,
            }));
        }
        catch {
            return null;
        }
    }
}
