// @vibehub/core/dist/index.js
export * from "./contract/map-types.js";
export * from "./contract/panel-types.js";
export * from "./contract/conflict-types.js";
export * from "./contract/install-types.js";
export * from "./contract/kb-types.js";
export * from "./contract/workbench-bridge.js";
export * from "./contract/workflow-receipt.js";
export { defaultDbPath, openDb, resolveDbPath, vibehubHome, CURRENT_SCHEMA_VERSION, inspectDatabase, withReadonlyDb, } from "./db.js";
export * from "./runtime-lifecycle.js";
export * from "./project-activation.js";
export * from "./activity-store.js";
// GraphStore is a read-model compatibility facade after the canonical KB cutover.
// Canonical/mapping writes are intentionally not part of the package API.
export { listTerritories, readTerritoryLayouts, countAnchoredFiles, featuresForFile, readSpec, edgesFrom, getSetting, } from "./graph-store.js";
export { GitFacade, GhFacade, GitError, parseRemoteSlug, } from "./git-facade.js";
export { getRepoByRoot, upsertRepo, readBranchFiles, readConflicts, readSyncState, readTeamBranches, } from "./team-store.js";
export { selectConflictCandidates, syncTeamSnapshot, } from "./team-sync.js";
export { exportTeamMapSnapshot } from "./snapshot-export.js";
export { RuntimeService, resolveWorkbenchRepoRef, } from "./runtime-service.js";
export { readTaskTimeline } from "./timeline-read.js";
export { canonicalConflictPair, isConflictPairIgnored, persistIgnoredConflictPair, resolveConflictPair, } from "./conflict-ignore.js";
export { readTaskPanelModel, readConflictDetailModel } from "./live-read-models.js";
export { applyIntervention, InterventionTargetNotFoundError, } from "./intervention-service.js";
export { squarify, layoutTerritories, DEFAULT_LAYOUT, } from "./treemap.js";
export { nextState } from "./state-machine.js";
export { classifyUserPrompt } from "./milestone.js";
export { canonicalRepoPath, claimOffScopeReminder, matchesScopePattern, readScopePatterns, replaceScopePatterns, } from "./scope-registry.js";
export * from "./knowledge-service.js";
export * from "./distillation-service.js";
export * from "./operation-dispatcher.js";
export * from "./operation-contracts.js";
export { GIT_SEMANTIC_STORE_RELATIVE_PATH, diffSemanticRefs, durableProvenanceId, exportGitSemanticStore, hasGitSemanticStore, inspectGitSemanticStoreWorktree, materializeSemanticCacheAtRef, materializeSemanticCacheFromWorktree, migrateSqliteSemanticStoreToGit, readSpecAtRef, replaceGitSemanticStore, stableSemanticPath, } from "./git-semantic-store.js";
export { ingestCanonicalHookEvent, ingestHookEvent, lastAssistantText, } from "./hook-ingest.js";
export { CHECKPOINT_CADENCE_SETTING_KEY, DEFAULT_CHECKPOINT_CADENCE_TURNS, formatCheckpointReminder, recordUserPromptTurn, readPromptReplayFacts, resolveCheckpointCadence, } from "./knowledge-checkpoint.js";
export * from "./workflow-receipt-projectors.js";
