// @vibehub/core/dist/contract/workflow-receipt-projection.js
import { validateWorkflowReceiptStructure } from "./workflow-receipt.js";
export function projectInjectionInterventionReceipt(input) {
    const canQueue = input.result.outcome === "applied" || input.result.outcome === "already_applied";
    if (canQueue && !validIds(input.result.injectionIds, true)) {
        throw new Error("queued intervention requires non-empty unique positive safe injection ids");
    }
    const expectedIds = input.intervention.kind === "inject_both" ? 2 : 1;
    if (canQueue && input.result.injectionIds.length !== expectedIds) {
        throw new Error(`${input.intervention.kind} requires exactly ${expectedIds} persisted injection ids`);
    }
    if (!canQueue && !validIds(input.result.injectionIds, false)) {
        throw new Error("intervention result contains invalid injection ids");
    }
    const outcome = canQueue
        ? "queued"
        : input.result.outcome === "unsupported"
            ? "failed"
            : "skipped";
    const receipt = {
        schemaVersion: 1,
        activity: "inject",
        phase: "complete",
        outcome,
        visibility: outcome === "failed" ? "expanded" : "brief",
        trigger: bounded(input.trigger),
        evidence: [{
                source: "applied_intervention",
                effect: "injection",
                outcome,
                subject: input.intervention.kind === "inject_both"
                    ? `${input.intervention.kind} conflict ${input.intervention.conflictId}`
                    : `${input.intervention.kind} task ${input.intervention.taskId}`,
                requestId: bounded(input.result.requestId),
                originalKind: input.intervention.kind,
                resultOutcome: input.result.outcome,
                ...(input.result.replayed === undefined ? {} : { replayed: input.result.replayed }),
                injectionIds: [...input.result.injectionIds],
                ...(input.result.message ? { detail: bounded(input.result.message) } : {}),
            }],
        nextAction: outcome === "failed"
            ? { required: true, instruction: bounded(input.result.message ?? "Use a supported injection intervention and retry.") }
            : null,
        at: input.result.acceptedAt,
    };
    const validation = validateWorkflowReceiptStructure(receipt);
    if (!validation.ok)
        throw new Error(`invalid WorkflowReceiptV1 projection: ${validation.errors.join("; ")}`);
    return receipt;
}
function validIds(value, nonEmpty) {
    return (!nonEmpty || value.length > 0)
        && value.every((id) => Number.isSafeInteger(id) && id > 0)
        && new Set(value).size === value.length;
}
function bounded(value) {
    const chars = [...value];
    return chars.length <= 20_000 ? value : `${chars.slice(0, 19_999).join("")}…`;
}
