// @vibehub/core/dist/knowledge-checkpoint.js
import { getSetting } from "./graph-store.js";
/**
 * Default user-turn cadence between checkpoint reminders. Tunable, awaits
 * dogfood benchmark — no empirical basis yet. Seed reasoning: the session
 * protocol already asks for ingest-at-decision-time, so the checkpoint is
 * a safety net; a handful of turns approximates one work phase without
 * nagging every exchange. Override per repo or globally via the settings
 * key below; invalid or non-positive values fall back here ("0" is not a
 * disable switch — disabling would be an explicit product decision).
 */
export const DEFAULT_CHECKPOINT_CADENCE_TURNS = 8;
/** Settings override key (repo row shadows the repo_id 0 global row). */
export const CHECKPOINT_CADENCE_SETTING_KEY = "checkpoint.cadence.user_turns";
/**
 * The one place cadence configuration is resolved; never throws. Accepted
 * grammar is an explicit positive decimal integer (surrounding whitespace
 * tolerated) — "1e3"/"0x10" style Number() leniency is not configuration.
 */
export function resolveCheckpointCadence(db, repoId) {
    const raw = getSetting(db, CHECKPOINT_CADENCE_SETTING_KEY, repoId);
    if (raw === null)
        return DEFAULT_CHECKPOINT_CADENCE_TURNS;
    const trimmed = raw.trim();
    if (!/^\d+$/.test(trimmed))
        return DEFAULT_CHECKPOINT_CADENCE_TURNS;
    const parsed = Number(trimmed);
    return Number.isSafeInteger(parsed) && parsed > 0
        ? parsed
        : DEFAULT_CHECKPOINT_CADENCE_TURNS;
}
/**
 * Record one UserPromptSubmit turn and decide whether the checkpoint
 * reminder fires. MUST be called inside the hook's open transaction, after
 * the task row is upserted (FKs). Rules, in order:
 * 1. mechanical reset — new provenance rows attributed to this task move
 *    last_write_turn to the current count;
 * 2. replay dedup — a prompt_id seen before never counts and never fires,
 *    so duplicated hook registrations are inert;
 * 3. strict priority — an eligible turn that is delivering claimed
 *    interventions defers (reminder turn untouched) and retries on the
 *    next quiet turn; a pause instruction is never contradicted.
 */
export function recordUserPromptTurn(db, input) {
    const { repoId, taskId, promptId, now } = input;
    const threshold = resolveCheckpointCadence(db, repoId);
    db.prepare(`INSERT INTO task_prompt_cadence (repo_id, task_id, provenance_high_water, updated_at)
     VALUES (?, ?, (SELECT COALESCE(MAX(id), 0) FROM kb_provenance_events WHERE repo_id = ? AND task_id = ?), ?)
     ON CONFLICT(repo_id, task_id) DO NOTHING`).run(repoId, taskId, repoId, taskId, now);
    const row = db.prepare(`SELECT counted_turns AS countedTurns, last_write_turn AS lastWriteTurn,
            last_reminder_turn AS lastReminderTurn, provenance_high_water AS provenanceHighWater
     FROM task_prompt_cadence WHERE repo_id = ? AND task_id = ?`).get(repoId, taskId);
    let { countedTurns, lastWriteTurn, lastReminderTurn, provenanceHighWater } = row;
    const maxProvenance = db.prepare(`SELECT COALESCE(MAX(id), 0) AS id FROM kb_provenance_events WHERE repo_id = ? AND task_id = ?`).get(repoId, taskId).id;
    if (maxProvenance > provenanceHighWater) {
        lastWriteTurn = countedTurns;
        provenanceHighWater = maxProvenance;
    }
    const isNew = db.prepare(`INSERT INTO task_prompt_seen (repo_id, task_id, prompt_id, seen_at)
     VALUES (?, ?, ?, ?)
     ON CONFLICT(task_id, prompt_id) DO NOTHING`).run(repoId, taskId, promptId, now).changes === 1;
    if (isNew)
        countedTurns += 1;
    const eligible = isNew && countedTurns - Math.max(lastWriteTurn, lastReminderTurn) >= threshold;
    const status = !isNew
        ? "duplicate"
        : eligible
            ? input.deliveringInterventions
                ? "deferred"
                : "fired"
            : "counted";
    if (status === "fired")
        lastReminderTurn = countedTurns;
    db.prepare(`UPDATE task_prompt_cadence
     SET counted_turns = ?, last_write_turn = ?, last_reminder_turn = ?,
         provenance_high_water = ?, updated_at = ?
     WHERE repo_id = ? AND task_id = ?`).run(countedTurns, lastWriteTurn, lastReminderTurn, provenanceHighWater, now, repoId, taskId);
    return {
        status,
        countedTurns,
        turnsSinceLastWrite: countedTurns - lastWriteTurn,
        threshold,
    };
}
/**
 * Read-only replay guard for the whole UserPromptSubmit hook boundary.
 * A known prompt must be rejected before task/session state, timeline, or
 * pending-injection claims can move.
 */
export function readPromptReplayFacts(db, input) {
    const seen = db.prepare(`SELECT 1 FROM task_prompt_seen
     WHERE repo_id = ? AND task_id = ? AND prompt_id = ?`).get(input.repoId, input.taskId, input.promptId);
    if (!seen)
        return null;
    const row = db.prepare(`SELECT counted_turns AS countedTurns, last_write_turn AS lastWriteTurn,
            provenance_high_water AS provenanceHighWater
     FROM task_prompt_cadence WHERE repo_id = ? AND task_id = ?`).get(input.repoId, input.taskId);
    if (!row)
        throw new Error(`missing checkpoint cadence for replay: ${input.taskId}`);
    const maxProvenance = db.prepare(`SELECT COALESCE(MAX(id), 0) AS id FROM kb_provenance_events
     WHERE repo_id = ? AND task_id = ?`).get(input.repoId, input.taskId).id;
    const effectiveLastWrite = maxProvenance > row.provenanceHighWater ? row.countedTurns : row.lastWriteTurn;
    return {
        status: "duplicate",
        countedTurns: row.countedTurns,
        turnsSinceLastWrite: row.countedTurns - effectiveLastWrite,
        threshold: resolveCheckpointCadence(db, input.repoId),
    };
}
/**
 * The injected reminder. It asks for judgment, never mandates a write —
 * "no durable knowledge" must stay a first-class outcome (no filler
 * records). The task id is embedded because the reset evidence is
 * task-attributed: a capture written without `--task <this id>` (the CLI
 * skill route) would never close the remind→capture→reset loop, and the
 * agent has no other way to learn the opaque id. Wording is a product
 * surface; changes need a spec trace.
 */
export function formatCheckpointReminder(facts, taskId) {
    const captured = facts.countedTurns > facts.turnsSinceLastWrite;
    const opening = captured
        ? `[Vibehub] Knowledge checkpoint: ${facts.turnsSinceLastWrite} user turns on this task since the last captured knowledge write. ` +
            "Review the conversation since then"
        : `[Vibehub] Knowledge checkpoint: ${facts.turnsSinceLastWrite} user turns on this task with no knowledge captured yet. ` +
            "Review the conversation";
    return (`${opening} for durable knowledge — new intent, decisions, constraints, contracts, conventions, context, or changes. ` +
        `If any exist, use the vibehub-ingest skill now and attribute the write to task ${taskId}. ` +
        "If none, continue working — do not create filler records.");
}
