// @vibehub/core/dist/team-store.js
export function upsertRepo(db, rootPath, slug, defaultBranch, now) {
    db.prepare(`INSERT INTO repos (root_path, slug, default_branch, created_at)
     VALUES (?, ?, ?, ?)
     ON CONFLICT(root_path) DO UPDATE SET
       slug = excluded.slug, default_branch = excluded.default_branch`).run(rootPath, slug, defaultBranch, now);
    return getRepoByRoot(db, rootPath);
}
export function getRepoByRoot(db, rootPath) {
    const row = db
        .prepare(`SELECT id, root_path AS rootPath, slug, default_branch AS defaultBranch
       FROM repos WHERE root_path = ?`)
        .get(rootPath);
    return row ?? null;
}
export function writeSyncState(db, repoId, s) {
    db.prepare(`INSERT INTO sync_state (repo_id, last_fetch_at, last_fetch_ok, gh_available, repo_files, last_synced_at)
     VALUES (?, ?, ?, ?, ?, ?)
     ON CONFLICT(repo_id) DO UPDATE SET
       last_fetch_at = excluded.last_fetch_at,
       last_fetch_ok = excluded.last_fetch_ok,
       gh_available = excluded.gh_available,
       repo_files = excluded.repo_files,
       last_synced_at = excluded.last_synced_at`).run(repoId, s.lastFetchAt, s.lastFetchOk === null ? null : s.lastFetchOk ? 1 : 0, s.ghAvailable ? 1 : 0, s.repoFiles, s.lastSyncedAt);
}
export function readSyncState(db, repoId) {
    const r = db
        .prepare(`SELECT last_fetch_at, last_fetch_ok, gh_available, repo_files, last_synced_at
       FROM sync_state WHERE repo_id = ?`)
        .get(repoId);
    if (!r)
        return null;
    return {
        lastFetchAt: r.last_fetch_at,
        lastFetchOk: r.last_fetch_ok === null ? null : r.last_fetch_ok === 1,
        ghAvailable: r.gh_available === 1,
        repoFiles: r.repo_files,
        lastSyncedAt: r.last_synced_at,
    };
}
/** Replace the branch set wholesale — the snapshot IS the remote state. */
export function replaceTeamBranches(db, repoId, branches) {
    const tx = db.transaction(() => {
        db.prepare(`DELETE FROM team_branches WHERE repo_id = ?`).run(repoId);
        const ins = db.prepare(`INSERT INTO team_branches
       (repo_id, name, head_sha, last_commit_at, last_author, ahead, behind, merged, pr_number, pr_state, pr_title)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`);
        for (const b of branches) {
            ins.run(repoId, b.name, b.headSha, b.lastCommitAt, b.lastAuthor, b.ahead, b.behind, b.merged ? 1 : 0, b.prNumber, b.prState, b.prTitle);
        }
    });
    tx();
}
export function readTeamBranches(db, repoId) {
    const rows = db
        .prepare(`SELECT name, head_sha AS headSha, last_commit_at AS lastCommitAt,
              last_author AS lastAuthor, ahead, behind, merged,
              pr_number AS prNumber, pr_state AS prState, pr_title AS prTitle
       FROM team_branches WHERE repo_id = ? ORDER BY last_commit_at DESC`)
        .all(repoId);
    // only the int→boolean column needs a JS conversion
    return rows.map((r) => ({ ...r, merged: r.merged === 1 }));
}
export function replaceBranchFiles(db, repoId, files) {
    const tx = db.transaction(() => {
        db.prepare(`DELETE FROM team_branch_files WHERE repo_id = ?`).run(repoId);
        const ins = db.prepare(`INSERT INTO team_branch_files (repo_id, branch, path, change_kind)
       VALUES (?, ?, ?, ?)`);
        for (const f of files)
            ins.run(repoId, f.branch, f.path, f.changeKind);
    });
    tx();
}
export function readBranchFiles(db, repoId, branch) {
    return db
        .prepare(`SELECT path, change_kind AS changeKind FROM team_branch_files
       WHERE repo_id = ? AND branch = ? ORDER BY path`)
        .all(repoId, branch);
}
/**
 * Upsert the conflict set for this sync. first_detected_at survives re-syncs
 * (the map's Conflict.detectedAt is "when first detected", not "last seen");
 * pairs that no longer conflict are pruned — resolved is resolved.
 */
export function reconcileConflicts(db, repoId, seen, now) {
    const tx = db.transaction(() => {
        const ins = db.prepare(`INSERT INTO team_conflicts (repo_id, branch_a, branch_b, path, first_detected_at, last_seen_at)
       VALUES (?, ?, ?, ?, ?, ?)
       ON CONFLICT(repo_id, branch_a, branch_b, path) DO UPDATE SET last_seen_at = excluded.last_seen_at`);
        for (const c of seen)
            ins.run(repoId, c.branchA, c.branchB, c.path, now, now);
        db.prepare(`DELETE FROM team_conflicts WHERE repo_id = ? AND last_seen_at != ?`).run(repoId, now);
    });
    tx();
}
export function readConflicts(db, repoId) {
    return db
        .prepare(`SELECT branch_a AS branchA, branch_b AS branchB, path,
              first_detected_at AS firstDetectedAt, last_seen_at AS lastSeenAt
       FROM team_conflicts WHERE repo_id = ?
       ORDER BY branch_a, branch_b, path`)
        .all(repoId);
}
