// @vibehub/core/dist/workflow-receipt-projectors.js
import { CANONICAL_OPERATION_PRESENTATION, validateWorkflowReceiptStructure, } from "./contract/workflow-receipt.js";
export const OPERATION_PRESENTATION = CANONICAL_OPERATION_PRESENTATION;
export function projectOperationReceipt(input) {
    const identity = operationIdentity(input);
    const presentation = presentationFor(identity.operation);
    const outcome = input.result.ok
        ? presentation.effect === "read" ? "returned" : "persisted"
        : "failed";
    const counts = input.result.ok && presentation.effect === "read"
        ? returnedCounts(input.result.data)
        : {};
    const nextAction = input.result.ok
        ? input.nextAction ?? null
        : {
            required: true,
            instruction: bounded(input.result.error.nextSafeActions[0]
                ?? `Inspect ${input.result.error.code} and retry the operation safely.`),
        };
    return checked({
        schemaVersion: 1,
        activity: presentation.activity,
        phase: "complete",
        outcome,
        visibility: input.result.ok ? input.visibility ?? "brief" : "expanded",
        trigger: bounded(input.trigger),
        evidence: [{
                source: "operation_result",
                operation: identity.operation,
                repoId: identity.repoId,
                requestId: identity.requestId,
                ok: input.result.ok,
                effect: presentation.effect,
                outcome,
                subject: `${identity.operation} request ${identity.requestId}`,
                ...(!input.result.ok
                    ? { detail: bounded(`${input.result.error.code}: ${input.result.error.message}`) }
                    : {}),
                ...counts,
            }],
        nextAction,
        at: identity.at,
    });
}
export function projectInitReceipt(input) {
    const outcome = input.result.ok ? "persisted" : "waiting";
    return checked({
        schemaVersion: 1,
        activity: "setup",
        phase: "complete",
        outcome,
        visibility: "expanded",
        trigger: bounded(input.trigger),
        evidence: [{
                source: "init_runtime_result",
                effect: "write",
                outcome,
                subject: bounded(input.result.repo.root),
                ok: input.result.ok,
                repoId: input.result.repo.id,
                schemaVersion: input.result.schemaVersion,
                conflictCount: input.result.conflicts.length,
            }],
        nextAction: input.result.ok
            ? null
            : { required: true, instruction: "Review managed asset conflicts before retrying setup." },
        at: input.at,
    });
}
export function projectDoctorReceipt(input) {
    const computedHealthy = input.result.db.status === "healthy"
        && input.result.nativeDependency.status === "healthy"
        && input.result.repo.status === "healthy"
        && input.result.managedAssets.status === "healthy";
    if (computedHealthy !== input.result.healthy) {
        throw new Error("doctor healthy boolean contradicts component statuses");
    }
    const outcome = computedHealthy ? "verified" : "failed";
    return checked({
        schemaVersion: 1,
        activity: "setup",
        phase: "complete",
        outcome,
        visibility: computedHealthy ? "brief" : "expanded",
        trigger: bounded(input.trigger),
        evidence: [{
                source: "doctor_runtime_result",
                effect: "health_check",
                outcome,
                subject: "VibeHub runtime connectivity",
                computedHealthy,
                dbStatus: input.result.db.status,
                nativeStatus: input.result.nativeDependency.status,
                repoStatus: input.result.repo.status,
                managedAssetsStatus: input.result.managedAssets.status,
            }],
        nextAction: computedHealthy
            ? null
            : { required: true, instruction: "Repair the unhealthy runtime checks before continuing setup." },
        at: input.at,
    });
}
// The applied-injection projection is browser-safe contract code so the App
// can consume the same receipt truth; re-exported here to keep the Node API
// surface unchanged.
export { projectInjectionInterventionReceipt, } from "./contract/workflow-receipt-projection.js";
export function projectInjectionClaimReceipt(input) {
    if (!validClaimed(input.claimed)) {
        throw new Error("claimed injections require complete mechanical results with unique positive safe injection ids");
    }
    return checked({
        schemaVersion: 1,
        activity: "inject",
        phase: "complete",
        outcome: "claimed",
        visibility: "brief",
        trigger: bounded(input.trigger),
        evidence: [{
                source: "hook_evidence",
                effect: "injection",
                outcome: "claimed",
                subject: `injection claim for task ${bounded(input.taskId)}`,
                hookEvent: bounded(input.hookEvent),
                injectionIds: input.claimed.map((item) => item.id),
                injectionModes: input.claimed.map((item) => item.mode),
            }],
        nextAction: null,
        at: input.at,
    });
}
/**
 * Projects the brief checkpoint receipt from the hook's deterministic
 * cadence facts. Below-threshold turns ("counted"/"duplicate") are
 * heartbeats with no consumer — they never project a receipt.
 */
export function projectCheckpointReceipt(input) {
    const { facts } = input;
    if (facts.status !== "fired" && facts.status !== "deferred") {
        throw new Error("below-threshold cadence turns do not project receipts");
    }
    const fired = facts.status === "fired";
    return checked({
        schemaVersion: 1,
        activity: "checkpoint",
        phase: fired ? "execute" : "prepare",
        outcome: fired ? "attempted" : "skipped",
        visibility: "brief",
        trigger: bounded(input.trigger),
        evidence: [{
                source: "checkpoint_hook",
                effect: "none",
                outcome: fired ? "attempted" : "skipped",
                subject: bounded(`knowledge checkpoint for task ${input.taskId}`),
                userTurnCount: facts.turnsSinceLastWrite,
                detail: fired
                    ? `${facts.countedTurns} turns counted; threshold ${facts.threshold}`
                    : `yielded to intervention delivery; threshold ${facts.threshold}`,
            }],
        nextAction: null,
        at: input.at,
    });
}
function operationIdentity(input) {
    if (input.result.ok) {
        if (input.attempt && (input.attempt.operation !== input.result.meta.operation
            || input.attempt.repoId !== input.result.meta.repoId
            || input.attempt.requestId !== input.result.meta.requestId)) {
            throw new Error("operation attempt identity does not match successful result meta");
        }
        return input.result.meta;
    }
    if (!input.attempt)
        throw new Error("failed operation projection requires attempt identity");
    return input.attempt;
}
function presentationFor(operation) {
    if (!Object.prototype.hasOwnProperty.call(OPERATION_PRESENTATION, operation)) {
        throw new Error(`unknown canonical operation: ${operation}`);
    }
    return OPERATION_PRESENTATION[operation];
}
function returnedCounts(data) {
    if (Array.isArray(data))
        return { returnedCount: data.length, totalCount: data.length };
    if (!isRecord(data))
        return {};
    let returnedCount;
    for (const key of ["items", "matches", "results", "rows", "features", "specs"]) {
        const value = data[key];
        if (Array.isArray(value)) {
            returnedCount = value.length;
            break;
        }
    }
    let totalCount;
    for (const key of ["total", "count"]) {
        const value = data[key];
        if (Number.isSafeInteger(value) && Number(value) >= 0) {
            totalCount = Number(value);
            break;
        }
    }
    if (returnedCount !== undefined && totalCount === undefined)
        totalCount = returnedCount;
    return {
        ...(returnedCount === undefined ? {} : { returnedCount }),
        ...(totalCount === undefined ? {} : { totalCount }),
    };
}
function checked(receipt) {
    const validation = validateWorkflowReceiptStructure(receipt);
    if (!validation.ok)
        throw new Error(`invalid WorkflowReceiptV1 projection: ${validation.errors.join("; ")}`);
    return receipt;
}
function validClaimed(value) {
    return value.length > 0
        && validIds(value.map((item) => item.id), true)
        && value.every((item) => (item.mode === "inject" || item.mode === "pause")
            && typeof item.text === "string"
            && (item.context === null || typeof item.context === "string")
            && typeof item.createdAt === "string"
            && item.createdAt.length > 0);
}
function validIds(value, nonEmpty) {
    return (!nonEmpty || value.length > 0)
        && value.every((id) => Number.isSafeInteger(id) && id > 0)
        && new Set(value).size === value.length;
}
function bounded(value) {
    const chars = [...value];
    return chars.length <= 20_000 ? value : `${chars.slice(0, 19_999).join("")}…`;
}
function isRecord(value) {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}
